plugins/spectate/spectate.cpp

```diff
@@ -0,0 +1,363 @@
+    config = World::GetPersistentData(CONFIG_KEY);
+
+    if (!config.isValid()) {
+        config = World::AddPersistentData(CONFIG_KEY);
+        saveConfig();
+    } else {
+        unpause_enabled = config.ival(UNPAUSE);
+        disengage_enabled = config.ival(DISENGAGE);
+        focus_jobs_enabled = config.ival(JOB_FOCUS);
+        tick_threshold = config.ival(TICK_THRESHOLD);
+        pause_lock->unlock();
+        enable_auto_unpause(out, unpause_enabled);
+    }
+    return DFHack::CR_OK;
+}
+
+DFhackCExport command_result plugin_enable(color_ostream &out, bool enable) {
+    namespace EM = EventManager;
+    if (enable && !enabled) {
+        out.print("Spectate mode enabled!\n");
+        using namespace EM::EventType;
+        EM::EventHandler ticking(onTick, 15);
+        EM::EventHandler start(onJobStart, 0);
+        EM::EventHandler complete(onJobCompletion, 0);
+        EM::registerListener(EventType::TICK, ticking, plugin_self);
+        EM::registerListener(EventType::JOB_STARTED, start, plugin_self);
+        EM::registerListener(EventType::JOB_COMPLETED, complete, plugin_self);
+        enabled = true; // enable_auto_unpause won't do anything without this set now
+        enable_auto_unpause(out, unpause_enabled);
+    } else if (!enable && enabled) {
+        // warp 8, engage!
+        out.print("Spectate mode disabled!\n");
+        EM::unregisterAll(plugin_self);
+        // we need to retain whether auto-unpause is enabled, but we also need to disable its effect
+        bool temp = unpause_enabled;
+        enable_auto_unpause(out, false);
+        unpause_enabled = temp;
+        job_tracker.clear();
+        freq.clear();
+    }
+    enabled = enable;
+    return DFHack::CR_OK;
+}
+
+DFhackCExport command_result plugin_onstatechange(color_ostream &out, state_change_event event) {
+    if (enabled) {
+        switch (event) {
+            case SC_MAP_UNLOADED:
+            case SC_BEGIN_UNLOAD:
+            case SC_WORLD_UNLOADED:
+                our_dorf = nullptr;
+                job_watched = nullptr;
+                following_dwarf = false;
+            default:
+                break;
+        }
+    }
+    return CR_OK;
+}
+
+DFhackCExport command_result plugin_onupdate(color_ostream &out) {
+    // keeps announcement pause settings locked
+    World::Update(); // from pause.h
+    if (lock_collision) {
+        if (unpause_enabled) {
+            // player asked for auto-unpause enabled
+            World::SaveAnnouncementSettings();
+            if (World::DisableAnnouncementPausing()) {
+                // now that we've got what we want, we can lock it down
+                lock_collision = false;
+            }
+        } else {
+            if (World::RestoreAnnouncementSettings()) {
+                lock_collision = false;
+            }
+        }
+    }
+    while (unpause_enabled && !world->status.popups.empty()) {
+        // dismiss announcement popup(s)
```

Comment:
```suggestion
    }
    int failsafe = 0;
    while (unpause_enabled && !world->status.popups.empty() && ++failsafe <= 10) {
        // dismiss announcement popup(s)
```

Reply:
@myk002
I thought I put this on the thread about the popup dismissal, but I guess not.